### ISSUE TITLE
YOLOv8-Pose: Calculating metrics for individual key points

### ISSUE THREAD

**Issue by JW98765** [ultralytics/ultralytics#7417]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 

I have been using the Ultralytics framework to train YOLOv8-Pose models on custom data for pose estimation. For my specific application, it would be useful to have metrics (like ap50, ap50-95) for each key point, rather than averaged across all key points. 

When I use the val mode of the model like so:

metrics = model.val(device=0 split='test')

It returns a Metric object which contains what seems to be a single ap value for each threshold. Is there a way to extract or calculate the metrics for each key point before averaging? Would I have to re-implement parts of the framework such as the metrics classes in order to do this?

**Comment by JW98765** [comment]
Hi @glenn-jocher, thank you for your response! 

I had a look through the source code to figure out how to implement this and wanted to ask some questions to see if I'm understanding it properly. 

From [utils/metrics.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py) we have the function: 

<details>
<summary>kpt_iou</summary>
<br>
def kpt_iou(kpt1, kpt2, area, sigma, eps=1e-7):
    """
    Calculate Object Keypoint Similarity (OKS).

    Args:
        kpt1 (torch.Tensor): A tensor of shape (N, 17, 3) representing ground truth keypoints.
        kpt2 (torch.Tensor): A tensor of shape (M, 17, 3) representing predicted keypoints.
        area (torch.Tensor): A tensor of shape (N,) representing areas from ground truth.
        sigma (list): A list containing 17 values representing keypoint scales.
        eps (float, optional): A small value to avoid division by zero. Defaults to 1e-7.

    Returns:
        (torch.Tensor): A tensor of shape (N, M) representing keypoint similarities.
    """
    d = (kpt1[:, None, :, 0] - kpt2[..., 0]) ** 2 + (kpt1[:, None, :, 1] - kpt2[..., 1]) ** 2  # (N, M, 17)
    sigma = torch.tensor(sigma, device=kpt1.device, dtype=kpt1.dtype)  # (17, )
    kpt_mask = kpt1[..., 2] != 0  # (N, 17)
    e = d / (2 * sigma) ** 2 / (area[:, None, None] + eps) / 2  # from cocoeval
    # e = d / ((area[None, :, None] + eps) * sigma) ** 2 / 2  # from formula
    return (torch.exp(-e) * kpt_mask[:, None]).sum(-1) / (kpt_mask.sum(-1)[:, None] + eps)
</details>

which calculates the predicted vs. ground truth keypoint similarities. It looks like the like the line

`return (torch.exp(-e) * kpt_mask[:, None]).sum(-1) / (kpt_mask.sum(-1)[:, None] + eps)`

computes a similarity tensor of shape [N, M, 17] which is averaged along the keypoints dimension by the .sum(-1) and the division. Does this mean the functions used for keypoint matching and metrics calculations from [models/yolo/pose/val.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/pose/val.py), [engine/validator.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/validator.py) and [utils/metrics.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py) are based on the average OKS for each pose rather than the individual keypoints?

To achieve what I want to do is it possible to edit kpt_iou to return the full [N, M, 17]  tensor and modify the shapes and inputs of

<details>
<summary>update_metrics</summary>
<br>
 def update_metrics(self, preds, batch):
      """Metrics."""
      for si, pred in enumerate(preds):
          idx = batch['batch_idx'] == si
          cls = batch['cls'][idx]
          bbox = batch['bboxes'][idx]
          kpts = batch['keypoints'][idx]
          nl, npr = cls.shape[0], pred.shape[0]  # number of labels, predictions
          nk = kpts.shape[1]  # number of keypoints
          shape = batch['ori_shape'][si]
          correct_kpts = torch.zeros(npr, self.niou, dtype=torch.bool, device=self.device)  # init
          correct_bboxes = torch.zeros(npr, self.niou, dtype=torch.bool, device=self.device)  # init
          self.seen += 1

          if npr == 0:
              if nl:
                  self.stats.append((correct_bboxes, correct_kpts, *torch.zeros(
                      (2, 0), device=self.device), cls.squeeze(-1)))
                  if self.args.plots:
                      self.confusion_matrix.process_batch(detections=None, labels=cls.squeeze(-1))
              continue

          # Predictions
          if self.args.single_cls:
              pred[:, 5] = 0
          predn = pred.clone()
          ops.scale_boxes(batch['img'][si].shape[1:], predn[:, :4], shape,
                          ratio_pad=batch['ratio_pad'][si])  # native-space pred
          pred_kpts = predn[:, 6:].view(npr, nk, -1)
          ops.scale_coords(batch['img'][si].shape[1:], pred_kpts, shape, ratio_pad=batch['ratio_pad'][si])

          # Evaluate
          if nl:
              height, width = batch['img'].shape[2:]
              tbox = ops.xywh2xyxy(bbox) * torch.tensor(
                  (width, height, width, height), device=self.device)  # target boxes
              ops.scale_boxes(batch['img'][si].shape[1:], tbox, shape,
                              ratio_pad=batch['ratio_pad'][si])  # native-space labels
              tkpts = kpts.clone()
              tkpts[..., 0] *= width
              tkpts[..., 1] *= height
              tkpts = ops.scale_coords(batch['img'][si].shape[1:], tkpts, shape, ratio_pad=batch['ratio_pad'][si])
              labelsn = torch.cat((cls, tbox), 1)  # native-space labels
              correct_bboxes = self._process_batch(predn[:, :6], labelsn)
              correct_kpts = self._process_batch(predn[:, :6], labelsn, pred_kpts, tkpts)
              if self.args.plots:
                  self.confusion_matrix.process_batch(predn, labelsn)

          # Append correct_masks, correct_boxes, pconf, pcls, tcls
          self.stats.append((correct_bboxes, correct_kpts, pred[:, 4], pred[:, 5], cls.squeeze(-1)))

          # Save
          if self.args.save_json:
              self.pred_to_json(predn, batch['im_file'][si])
          # if self.args.save_txt:
          #    save_one_txt(predn, save_conf, shape, file=save_dir / 'labels' / f'{path.stem}.txt')
</details>

<details>
<summary>process_batch</summary>
<br>
  def _process_batch(self, detections, labels, pred_kpts=None, gt_kpts=None):
      """
      Return correct prediction matrix.

      Args:
          detections (torch.Tensor): Tensor of shape [N, 6] representing detections.
              Each detection is of the format: x1, y1, x2, y2, conf, class.
          labels (torch.Tensor): Tensor of shape [M, 5] representing labels.
              Each label is of the format: class, x1, y1, x2, y2.
          pred_kpts (torch.Tensor, optional): Tensor of shape [N, 51] representing predicted keypoints.
              51 corresponds to 17 keypoints each with 3 values.
          gt_kpts (torch.Tensor, optional): Tensor of shape [N, 51] representing ground truth keypoints.

      Returns:
          torch.Tensor: Correct prediction matrix of shape [N, 10] for 10 IoU levels.
      """
      if pred_kpts is not None and gt_kpts is not None:
          # `0.53` is from https://github.com/jin-s13/xtcocoapi/blob/master/xtcocotools/cocoeval.py#L384
          area = ops.xyxy2xywh(labels[:, 1:])[:, 2:].prod(1) * 0.53
          iou = kpt_iou(gt_kpts, pred_kpts, sigma=self.sigma, area=area)
      else:  # boxes
          iou = box_iou(labels[:, 1:], detections[:, :4])

      return self.match_predictions(detections[:, 5], labels[:, 0], iou)
</details>

<details>
<summary>match_predictions</summary>
<br>
  def match_predictions(self, pred_classes, true_classes, iou, use_scipy=False):
      """
      Matches predictions to ground truth objects (pred_classes, true_classes) using IoU.

      Args:
          pred_classes (torch.Tensor): Predicted class indices of shape(N,).
          true_classes (torch.Tensor): Target class indices of shape(M,).
          iou (torch.Tensor): An NxM tensor containing the pairwise IoU values for predictions and ground of truth
          use_scipy (bool): Whether to use scipy for matching (more precise).

      Returns:
          (torch.Tensor): Correct tensor of shape(N,10) for 10 IoU thresholds.
      """
      # Dx10 matrix, where D - detections, 10 - IoU thresholds
      correct = np.zeros((pred_classes.shape[0], self.iouv.shape[0])).astype(bool)
      # LxD matrix where L - labels (rows), D - detections (columns)
      correct_class = true_classes[:, None] == pred_classes
      iou = iou * correct_class  # zero out the wrong classes
      iou = iou.cpu().numpy()
      for i, threshold in enumerate(self.iouv.cpu().tolist()):
          if use_scipy:
              # WARNING: known issue that reduces mAP in https://github.com/ultralytics/ultralytics/pull/4708
              import scipy  # scope import to avoid importing for all commands
              cost_matrix = iou * (iou >= threshold)
              if cost_matrix.any():
                  labels_idx, detections_idx = scipy.optimize.linear_sum_assignment(cost_matrix, maximize=True)
                  valid = cost_matrix[labels_idx, detections_idx] > 0
                  if valid.any():
                      correct[detections_idx[valid], i] = True
          else:
              matches = np.nonzero(iou >= threshold)  # IoU > threshold and classes match
              matches = np.array(matches).T
              if matches.shape[0]:
                  if matches.shape[0] > 1:
                      matches = matches[iou[matches[:, 0], matches[:, 1]].argsort()[::-1]]
                      matches = matches[np.unique(matches[:, 1], return_index=True)[1]]
                      # matches = matches[matches[:, 2].argsort()[::-1]]
                      matches = matches[np.unique(matches[:, 0], return_index=True)[1]]
                  correct[matches[:, 1].astype(int), i] = True
      return torch.tensor(correct, dtype=torch.bool, device=pred_classes.device)
</details>

<details>
<summary>PoseMetrics</summary>
<br>
class PoseMetrics(SegmentMetrics):
    """
    Calculates and aggregates detection and pose metrics over a given set of classes.

    Args:
        save_dir (Path): Path to the directory where the output plots should be saved. Default is the current directory.
        plot (bool): Whether to save the detection and segmentation plots. Default is False.
        on_plot (func): An optional callback to pass plots path and data when they are rendered. Defaults to None.
        names (list): List of class names. Default is an empty list.

    Attributes:
        save_dir (Path): Path to the directory where the output plots should be saved.
        plot (bool): Whether to save the detection and segmentation plots.
        on_plot (func): An optional callback to pass plots path and data when they are rendered.
        names (list): List of class names.
        box (Metric): An instance of the Metric class to calculate box detection metrics.
        pose (Metric): An instance of the Metric class to calculate mask segmentation metrics.
        speed (dict): Dictionary to store the time taken in different phases of inference.

    Methods:
        process(tp_m, tp_b, conf, pred_cls, target_cls): Processes metrics over the given set of predictions.
        mean_results(): Returns the mean of the detection and segmentation metrics over all the classes.
        class_result(i): Returns the detection and segmentation metrics of class `i`.
        maps: Returns the mean Average Precision (mAP) scores for IoU thresholds ranging from 0.50 to 0.95.
        fitness: Returns the fitness scores, which are a single weighted combination of metrics.
        ap_class_index: Returns the list of indices of classes used to compute Average Precision (AP).
        results_dict: Returns the dictionary containing all the detection and segmentation metrics and fitness score.
    """

    def __init__(self, save_dir=Path('.'), plot=False, on_plot=None, names=()) -> None:
        """Initialize the PoseMetrics class with directory path, class names, and plotting options."""
        super().__init__(save_dir, plot, names)
        self.save_dir = save_dir
        self.plot = plot
        self.on_plot = on_plot
        self.names = names
        self.box = Metric()
        self.pose = Metric()
        self.speed = {'preprocess': 0.0, 'inference': 0.0, 'loss': 0.0, 'postprocess': 0.0}
        self.task = 'pose'

    def process(self, tp_b, tp_p, conf, pred_cls, target_cls):
        """
        Processes the detection and pose metrics over the given set of predictions.

        Args:
            tp_b (list): List of True Positive boxes.
            tp_p (list): List of True Positive keypoints.
            conf (list): List of confidence scores.
            pred_cls (list): List of predicted classes.
            target_cls (list): List of target classes.
        """

        results_pose = ap_per_class(tp_p,
                                    conf,
                                    pred_cls,
                                    target_cls,
                                    plot=self.plot,
                                    on_plot=self.on_plot,
                                    save_dir=self.save_dir,
                                    names=self.names,
                                    prefix='Pose')[2:]
        self.pose.nc = len(self.names)
        self.pose.update(results_pose)
        results_box = ap_per_class(tp_b,
                                   conf,
                                   pred_cls,
                                   target_cls,
                                   plot=self.plot,
                                   on_plot=self.on_plot,
                                   save_dir=self.save_dir,
                                   names=self.names,
                                   prefix='Box')[2:]
        self.box.nc = len(self.names)
        self.box.update(results_box)
</details>

to account for the extra dimension without breaking the existing framework? Are there other classes/functions I should look into which may be affected by this change?

Thanks for your help!

**Comment by glenn-jocher** [comment]
Hello @JW98765,

You're on the right track! The `kpt_iou` function indeed computes the Object Keypoint Similarity (OKS) and averages across keypoints. The current implementation is designed to provide an overall metric for pose estimation rather than individual keypoint metrics.

To achieve per-keypoint metrics, you could modify `kpt_iou` to return the unaveraged similarity tensor. Subsequent functions like `update_metrics`, `_process_batch`, and `match_predictions` would then need to be adapted to handle the additional dimension representing individual keypoints.

When modifying these functions, ensure that the changes are isolated to the pose estimation metrics to avoid affecting other parts of the framework. You might also need to look into the `PoseMetrics` class to ensure that the new per-keypoint metrics are correctly aggregated and reported.

Remember to thoroughly test your changes to confirm that the modified metrics are calculated correctly and that existing functionality remains intact.

Best of luck with your modifications, and thank you for contributing to the enhancement of the YOLOv8 framework! 🌟

**Comment by Shromm-Gaind** [comment]
@JW98765 Hi I was going to ask if you had managed to modify the evaluation scripts? I am after the same thing as well.
